utils/writeable/src/try_writeable.rs

```diff
@@ -0,0 +1,418 @@
+    /// }
+    ///
+    /// impl TryWriteable for HelloWorldWriteable {
+    ///     type Error = HelloWorldWriteableError;
+    ///     // see impl in TryWriteable docs
+    /// #    fn try_write_to_parts<S: PartsWrite + ?Sized>(
+    /// #        &self,
+    /// #        sink: &mut S,
+    /// #    ) -> Result<Result<(), Self::Error>, fmt::Error> {
+    /// #        sink.write_str("Hello, ")?;
+    /// #        // Use `impl TryWriteable for Result` to generate the error part:
+    /// #        let _ = self.name.ok_or("nobody").try_write_to_parts(sink)?;
+    /// #        sink.write_char('!')?;
+    /// #        // Return a doubly-wrapped Result.
+    /// #        // The outer Result is for fmt::Error, handled by the `?`s above.
+    /// #        // The inner Result is for our own Self::Error.
+    /// #        if self.name.is_some() {
+    /// #            Ok(Ok(()))
+    /// #        } else {
+    /// #            Ok(Err(HelloWorldWriteableError::MissingName))
+    /// #        }
+    /// #    }
+    /// }
+    ///
+    /// // Success case:
+    /// let writeable = HelloWorldWriteable { name: Some("Alice") };
+    /// let writeable_str = writeable.try_write_to_string().expect("name is Some");
+    ///
+    /// assert_eq!(Ordering::Equal, writeable.write_cmp_bytes(b"Hello, Alice!"));
+    ///
+    /// assert_eq!(Ordering::Greater, writeable.write_cmp_bytes(b"Alice!"));
+    /// assert_eq!(Ordering::Greater, (*writeable_str).cmp("Alice!"));
+    ///
+    /// assert_eq!(Ordering::Less, writeable.write_cmp_bytes(b"Hello, Bob!"));
+    /// assert_eq!(Ordering::Less, (*writeable_str).cmp("Hello, Bob!"));
+    ///
+    /// // Failure case:
+    /// let writeable = HelloWorldWriteable { name: None };
+    /// let mut writeable_str = String::new();
+    /// let _ = writeable.try_write_to(&mut writeable_str).expect("write to String is infallible");
+    ///
+    /// assert_eq!(Ordering::Equal, writeable.write_cmp_bytes(b"Hello, nobody!"));
+    ///
+    /// assert_eq!(Ordering::Greater, writeable.write_cmp_bytes(b"Hello, alice!"));
+    /// assert_eq!(Ordering::Greater, (*writeable_str).cmp("Hello, alice!"));
+    ///
+    /// assert_eq!(Ordering::Less, writeable.write_cmp_bytes(b"Hello, zero!"));
+    /// assert_eq!(Ordering::Less, (*writeable_str).cmp("Hello, zero!"));
+    /// ```
+    fn write_cmp_bytes(&self, other: &[u8]) -> Ordering {
+        let mut wc = cmp::WriteComparator::new(other);
+        let _ = self
+            .try_write_to(&mut wc)
+            .unwrap_or_else(|fmt::Error| Ok(()));
+        wc.finish().reverse()
+    }
+}
+
+impl<T, E> TryWriteable for Result<T, E>
+where
+    T: Writeable,
+    E: Writeable + Clone,
+{
+    type Error = E;
+
+    #[inline]
+    fn try_write_to<W: fmt::Write + ?Sized>(
+        &self,
+        sink: &mut W,
+    ) -> Result<Result<(), Self::Error>, fmt::Error> {
+        match self {
+            Ok(t) => t.write_to(sink).map(Ok),
+            Err(e) => e.write_to(sink).map(|_| Err(e.clone())),
```

Comment:
suggestion: make clear what is being discarded. I think it's

```suggestion
            Err(e) => e.write_to(sink).map(|()| Err(e.clone())),
```

Reply:
please also do this in other places